accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,577 @@
+
+_For naming, also consider `S.T.J.DynamicSerialization.dll`._
+
+A more open-ended name of `S.T.J.Extensions.dll` was considered but is not discoverable, so a more feature-specific name was considered a better choice. By going with feature-specific names like `S.T.J.Dynamic.dll`, we may end up with additional assemblies in the future. If the need arises, these extension assemblies could forward to private common implementation assembly(s).
+
+## Proposed API to enable dynamic mode
+In the new STJD assembly:
+```cs
+namespace System.Text.Json
+{
+    public static class JsonExtensions
+    {
+        public static void EnableDynamic(JsonSerializerOptions this options);
+    }
+}
+```
+
+Thus a consumer would reference STJD.dll and then author code such as:
+```cs
+    var options = new JsonSerializerOptions();
+    options.EnableDynamicTypes();
+
+    dynamic obj = JsonSerializer.Deserialize<dynamic>("{\"MyProp\":42}", options);
+    int i = obj.MyProp;
+    Debug.Assert(i == 42);
+
+    string json = JsonSerializer.Serialize(obj);
+```
+
+### Internal implementation
+The types that are used at runtime to implement `dynamic` can be made `internal`. Instances are created by the System.Object custom converter which understands `dynamic`.
+```cs
+internal class JsonDynamicObject : JsonObject, IDynamicMetaObjectProvider {}
+internal class JsonDynamicArray : JsonArray, IDynamicMetaObjectProvider {}
+internal class JsonDynamicValue : JsonValue, IDynamicMetaObjectProvider {}
+```
+
+where `JsonObject`, `JsonArray`, and `JsonValue` are new public types in STJ.dll.
+
+### Varying the `T` in `Deserialize<T>`
+This
+```cs
+    dynamic obj = JsonSerializer.Deserialize<dynamic>("{\"MyProp\":42}", options);
+```
+is equivalent to
+```cs
+    dynamic obj = JsonSerializer.Deserialize<object>("{\"MyProp\":42}", options);
+```
+and these (assuming `options.EnableDynamicTypes()` was called)
+```cs
+    dynamic obj = JsonSerializer.Deserialize<JsonNode>("{\"MyProp\":42}", options);
+    dynamic obj = JsonSerializer.Deserialize<JsonObject>("{\"MyProp\":42}", options);
+```
+
+### Supporting non-dynamic to dynamic
+If a previously created non-dynamic `JsonNode` object needs dynamic support, an extension could be provided:
+```cs
+namespace System.Text.Json
+{
+    public static class JsonExtensions
+    {
+        public static dynamic ToDynamic(JsonNode this value);
+    }
+}
+```
+
+Called like this:
+```cs
+    JsonNode node = ... // a non-dynamic JsonNode
+    dynamic obj = node.ToDynamic();
+```
+
+The implementation of `ToDynamic()` is somewhat slow since it does a deep clone. It would be similar to the code below except a bit faster since some overhead of the serialize\deserialize methods may be avoided:
```

Comment:
>The implementation of `ToDynamic()` is somewhat slow since it does a deep clone.

Does it have to do a deep clone?

Reply:
The current design has the dynamic type deriving from a non-dynamic base class. The base class contains the state.

When `ToDynamic` is called it must do a clone of some sort:
- If a shallow clone, the dynamic instance returned would also mutate the original non-dynamic instance. Plus the state itself would need to be kept in a helper class which complicates things a bit and would need to be public since the dynamic class is currently in a different assembly (pending possible linker work in 6.0).
- If a deep clone there are no issues with shared mutability and possible side-effects from that. (yes it's slow).

I don't anticipate `ToDynamic()` being a common scenario, however.